File: src/main/java/org/academy/api/client/gui/render/ScissorRect.java
package org.academy.api.client.gui.render;

import net.minecraft.util.Mth;
import org.joml.Matrix3x2f;
import org.joml.Vector2f;
import org.jspecify.annotations.Nullable;

import java.util.Objects;

import static org.academy.api.common.util.MathUtil.Axis2D;
import static org.academy.api.common.util.MathUtil.Direction2D;

public final class ScissorRect implements Comparable<ScissorRect> {
    private static final ScissorRect EMPTY = new ScissorRect(0, 0, 0, 0);

    private final Position2D position;
    private final float width;
    private final float height;

    public ScissorRect(Position2D position, float width, float height) {
        this.position = position;
        this.width = width;
        this.height = height;
    }

    public ScissorRect(float x, float y, float width, float height) {
        this(new Position2D(x, y), width, height);
    }

    public static ScissorRect empty() {
        return EMPTY;
    }

    public static ScissorRect of(Axis2D pAxis, float pPrimaryPosition, float pSecondaryPosition, float pPrimaryLength, float pSecondaryLength) {
        return switch (pAxis) {
            case HORIZONTAL -> new ScissorRect(pPrimaryPosition, pSecondaryPosition, pPrimaryLength, pSecondaryLength);
            case VERTICAL -> new ScissorRect(pSecondaryPosition, pPrimaryPosition, pSecondaryLength, pPrimaryLength);
        };
    }

    public Position2D getPosition() {
        return position;
    }

    public float getWidth() {
        return width;
    }

    public float getHeight() {
        return height;
    }

    public ScissorRect step(Direction2D pDirection) {
        return new ScissorRect(position.step(pDirection), width, height);
    }

    public float getLength(Axis2D axis) {
        return switch (axis) {
            case HORIZONTAL -> width;
            case VERTICAL -> height;
        };
    }

    public float getBoundInDirection(Direction2D pDirection) {
        var screenaxis = pDirection.getAxis();
        return pDirection.isPositive() ? position.getCoordinate(screenaxis) + getLength(screenaxis) - 1 : position.getCoordinate(screenaxis);
    }

    public ScissorRect getBorder(Direction2D pDirection) {
        var i = getBoundInDirection(pDirection);
        var screenaxis = pDirection.getAxis().orthogonal();
        var j = getBoundInDirection(screenaxis.getNegative());
        var k = getLength(screenaxis);
        return of(pDirection.getAxis(), i, j, 1, k).step(pDirection);
    }

    public boolean overlaps(ScissorRect pRectangle) {
        return overlapsInAxis(pRectangle, Axis2D.HORIZONTAL) && overlapsInAxis(pRectangle, Axis2D.VERTICAL);
    }

    public boolean overlapsInAxis(ScissorRect pRectangle, Axis2D pAxis) {
        var i = getBoundInDirection(pAxis.getNegative());
        var j = pRectangle.getBoundInDirection(pAxis.getNegative());
        var k = getBoundInDirection(pAxis.getPositive());
        var l = pRectangle.getBoundInDirection(pAxis.getPositive());
        return Math.max(i, j) <= Math.min(k, l);
    }

    public float getCenterInAxis(Axis2D pAxis) {
        return (getBoundInDirection(pAxis.getPositive()) + getBoundInDirection(pAxis.getNegative())) / 2;
    }

    @Nullable
    public ScissorRect intersection(ScissorRect pRectangle) {
        var i = Math.max(getLeft(), pRectangle.getLeft());
        var j = Math.max(getTop(), pRectangle.getTop());
        var k = Math.min(getRight(), pRectangle.getRight());
        var l = Math.min(getBottom(), pRectangle.getBottom());
        return i < k && j < l ? new ScissorRect(i, j, k - i, l - j) : null;
    }

    public boolean intersects(ScissorRect pRectangle) {
        return getLeft() < pRectangle.getRight()
                && getRight() > pRectangle.getLeft()
                && getTop() < pRectangle.getBottom()
                && getBottom() > pRectangle.getTop();
    }

    public boolean encompasses(ScissorRect pRectangle) {
        return pRectangle.getLeft() >= getLeft()
                && pRectangle.getTop() >= getTop()
                && pRectangle.getRight() <= getRight()
                && pRectangle.getBottom() <= getBottom();
    }

    public float getTop() {
        return position.y;
    }

    public float getBottom() {
        return position.y + height;
    }

    public float getLeft() {
        return position.x;
    }

    public float getRight() {
        return position.x + width;
    }

    public boolean containsPoint(int pX, int pY) {
        return pX >= getLeft() && pX < getRight() && pY >= getTop() && pY < getBottom();
    }

    public ScissorRect transformAxisAligned(Matrix3x2f pPos) {
        var vector2f = pPos.transformPosition(getLeft(), getTop(), new Vector2f());
        var vector2f1 = pPos.transformPosition(getRight(), getBottom(), new Vector2f());
        return new ScissorRect(
                Mth.floor(vector2f.x), Mth.floor(vector2f.y), Mth.floor(vector2f1.x - vector2f.x), Mth.floor(vector2f1.y - vector2f.y)
        );
    }

    public ScissorRect transformMaxBounds(Matrix3x2f pPos) {
        var vector2f = pPos.transformPosition(getLeft(), getTop(), new Vector2f());
        var vector2f1 = pPos.transformPosition(getRight(), getTop(), new Vector2f());
        var vector2f2 = pPos.transformPosition(getLeft(), getBottom(), new Vector2f());
        var vector2f3 = pPos.transformPosition(getRight(), getBottom(), new Vector2f());
        var f = Math.min(Math.min(vector2f.x(), vector2f2.x()), Math.min(vector2f1.x(), vector2f3.x()));
        var f1 = Math.max(Math.max(vector2f.x(), vector2f2.x()), Math.max(vector2f1.x(), vector2f3.x()));
        var f2 = Math.min(Math.min(vector2f.y(), vector2f2.y()), Math.min(vector2f1.y(), vector2f3.y()));
        var f3 = Math.max(Math.max(vector2f.y(), vector2f2.y()), Math.max(vector2f1.y(), vector2f3.y()));
        return new ScissorRect(Mth.floor(f), Mth.floor(f2), Mth.ceil(f1 - f), Mth.ceil(f3 - f2));
    }

    @Override
    public int compareTo(ScissorRect other) {
        var cmp = Float.compare(position.y, other.position.y);
        if (cmp != 0)
            return cmp;
        cmp = Float.compare(position.x, other.position.x);
        if (cmp != 0)
            return cmp;
        cmp = Float.compare(height, other.height);
        if (cmp != 0)
            return cmp;
        return Float.compare(width, other.width);
    }

    @Override
    public boolean equals(Object o) {
        if (this == o)
            return true;
        if (o == null || getClass() != o.getClass())
            return false;
        var that = (ScissorRect) o;
        return Float.compare(that.width, width) == 0
                && Float.compare(that.height, height) == 0
                && Objects.equals(position, that.position);
    }

    @Override
    public int hashCode() {
        return Objects.hash(position, width, height);
    }

    @Override
    public String toString() {
        return "ScissorRect[" + "position=" + position + ", " + "width=" + width + ", " + "height=" + height + ']';
    }
}

File: src/main/kotlin/org/academy/api/client/gui/render/ScissorRect.kt
package org.academy.api.client.gui.render

import net.minecraft.util.Mth
import org.academy.api.common.util.MathUtil.Axis2D
import org.academy.api.common.util.MathUtil.Direction2D
import org.joml.Matrix3x2f
import org.joml.Vector2f
import java.util.*
import kotlin.math.max
import kotlin.math.min

class ScissorRect(val position: Position2D, val width: Float, val height: Float) : Comparable<ScissorRect> {
    constructor(x: Float, y: Float, width: Float, height: Float) : this(Position2D(x, y), width, height)

    fun step(pDirection: Direction2D): ScissorRect {
        return ScissorRect(position.step(pDirection), width, height)
    }

    fun getLength(axis: Axis2D): Float {
        return when (axis) {
            Axis2D.HORIZONTAL -> width
            Axis2D.VERTICAL -> height
        }
    }

    fun getBoundInDirection(pDirection: Direction2D): Float {
        val screenaxis = pDirection.axis
        return if (pDirection.isPositive) position.getCoordinate(screenaxis) + getLength(screenaxis) - 1 else position.getCoordinate(
            screenaxis
        )
    }

    fun getBorder(pDirection: Direction2D): ScissorRect {
        val i = getBoundInDirection(pDirection)
        val screenaxis = pDirection.axis.orthogonal()
        val j = getBoundInDirection(screenaxis.negative)
        val k = getLength(screenaxis)
        return of(pDirection.axis, i, j, 1f, k).step(pDirection)
    }

    fun overlaps(pRectangle: ScissorRect): Boolean {
        return overlapsInAxis(pRectangle, Axis2D.HORIZONTAL) && overlapsInAxis(pRectangle, Axis2D.VERTICAL)
    }

    fun overlapsInAxis(pRectangle: ScissorRect, pAxis: Axis2D): Boolean {
        val i = getBoundInDirection(pAxis.negative)
        val j = pRectangle.getBoundInDirection(pAxis.negative)
        val k = getBoundInDirection(pAxis.positive)
        val l = pRectangle.getBoundInDirection(pAxis.positive)
        return max(i, j) <= min(k, l)
    }

    fun getCenterInAxis(pAxis: Axis2D): Float {
        return (getBoundInDirection(pAxis.positive) + getBoundInDirection(pAxis.negative)) / 2
    }

    fun intersection(pRectangle: ScissorRect): ScissorRect? {
        val i = max(this.left, pRectangle.left)
        val j = max(this.top, pRectangle.top)
        val k = min(this.right, pRectangle.right)
        val l = min(this.bottom, pRectangle.bottom)
        return if (i < k && j < l) ScissorRect(i, j, k - i, l - j) else null
    }

    fun intersects(pRectangle: ScissorRect): Boolean {
        return this.left < pRectangle.right && this.right > pRectangle.left && this.top < pRectangle.bottom && this.bottom > pRectangle.top
    }

    fun encompasses(pRectangle: ScissorRect): Boolean {
        return pRectangle.left >= this.left && pRectangle.top >= this.top && pRectangle.right <= this.right && pRectangle.bottom <= this.bottom
    }

    val top: Float
        get() = position.y

    val bottom: Float
        get() = position.y + height

    val left: Float
        get() = position.x

    val right: Float
        get() = position.x + width

    fun containsPoint(pX: Int, pY: Int): Boolean {
        return pX >= this.left && pX < this.right && pY >= this.top && pY < this.bottom
    }

    fun transformAxisAligned(pPos: Matrix3x2f): ScissorRect {
        val vector2f = pPos.transformPosition(this.left, this.top, Vector2f())
        val vector2f1 = pPos.transformPosition(this.right, this.bottom, Vector2f())
        return ScissorRect(
            Mth.floor(vector2f.x).toFloat(),
            Mth.floor(vector2f.y).toFloat(),
            Mth.floor(vector2f1.x - vector2f.x).toFloat(),
            Mth.floor(vector2f1.y - vector2f.y).toFloat()
        )
    }

    fun transformMaxBounds(pPos: Matrix3x2f): ScissorRect {
        val vector2f = pPos.transformPosition(this.left, this.top, Vector2f())
        val vector2f1 = pPos.transformPosition(this.right, this.top, Vector2f())
        val vector2f2 = pPos.transformPosition(this.left, this.bottom, Vector2f())
        val vector2f3 = pPos.transformPosition(this.right, this.bottom, Vector2f())
        val f = min(min(vector2f.x(), vector2f2.x()), min(vector2f1.x(), vector2f3.x()))
        val f1 = max(max(vector2f.x(), vector2f2.x()), max(vector2f1.x(), vector2f3.x()))
        val f2 = min(min(vector2f.y(), vector2f2.y()), min(vector2f1.y(), vector2f3.y()))
        val f3 = max(max(vector2f.y(), vector2f2.y()), max(vector2f1.y(), vector2f3.y()))
        return ScissorRect(
            Mth.floor(f).toFloat(),
            Mth.floor(f2).toFloat(),
            Mth.ceil(f1 - f).toFloat(),
            Mth.ceil(f3 - f2).toFloat()
        )
    }

    override fun compareTo(other: ScissorRect): Int {
        var cmp = position.y.compareTo(other.position.y)
        if (cmp != 0) return cmp
        cmp = position.x.compareTo(other.position.x)
        if (cmp != 0) return cmp
        cmp = height.compareTo(other.height)
        if (cmp != 0) return cmp
        return width.compareTo(other.width)
    }

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (other == null || javaClass != other.javaClass) return false
        val that = other as ScissorRect
        return that.width.compareTo(width) == 0 && that.height.compareTo(height) == 0 && position == that.position
    }

    override fun hashCode(): Int {
        return Objects.hash(position, width, height)
    }

    override fun toString(): String {
        return "ScissorRect[position=$position, width=$width, height=$height]"
    }

    companion object {
        private val EMPTY = ScissorRect(0f, 0f, 0f, 0f)

        fun empty(): ScissorRect {
            return EMPTY
        }

        fun of(
            pAxis: Axis2D,
            pPrimaryPosition: Float,
            pSecondaryPosition: Float,
            pPrimaryLength: Float,
            pSecondaryLength: Float
        ): ScissorRect {
            return when (pAxis) {
                Axis2D.HORIZONTAL -> ScissorRect(pPrimaryPosition, pSecondaryPosition, pPrimaryLength, pSecondaryLength)
                Axis2D.VERTICAL -> ScissorRect(pSecondaryPosition, pPrimaryPosition, pSecondaryLength, pPrimaryLength)
            }
        }
    }
}